File: app/src/main/java/com/example/macarus0/walkiewalkie/viewmodel/WalkieViewModel.java
package com.example.macarus0.walkiewalkie.viewmodel;

import android.app.Application;
import android.arch.lifecycle.AndroidViewModel;
import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.MutableLiveData;
import android.content.Context;
import android.support.annotation.NonNull;
import android.util.Log;

import com.example.macarus0.walkiewalkie.data.Dog;
import com.example.macarus0.walkiewalkie.data.DogOwner;
import com.example.macarus0.walkiewalkie.data.Owner;
import com.example.macarus0.walkiewalkie.data.Walk;
import com.example.macarus0.walkiewalkie.data.WalkLocation;
import com.example.macarus0.walkiewalkie.data.WalkPhoto;
import com.example.macarus0.walkiewalkie.data.WalkWithDogs;
import com.example.macarus0.walkiewalkie.data.WalkieDatabase;
import com.example.macarus0.walkiewalkie.data.WalkieDatabaseProvider;

import java.util.ArrayList;
import java.util.List;

public class WalkieViewModel extends AndroidViewModel {

    private final Context applicationContext;
    private WalkieDatabase mDb;

    private String TAG = "WalkieViewModel";

    public WalkieViewModel(@NonNull Application application) {
        super(application);
        applicationContext = application.getApplicationContext();
    }

    private WalkieDatabase getDb() {
        if (mDb == null) {
            mDb = WalkieDatabaseProvider.getDatabase(applicationContext);
        }
        return mDb;
    }

    // Methods used in overall views
    public LiveData<List<Dog>> getAllDogs() {
        return getDb().getDogDao().getAllDogs();
    }

    public LiveData<List<Owner>> getAllOwners() {
        return getDb().getOwnerDao().getAllOwners();
    }

    public LiveData<List<Walk>> getAllWalks() {
        return getDb().getWalkDao().getAllWalks();
    }

    // Methods used in Owner Views

    public LiveData<List<Dog>> getAllAvailableDogs(long ownerId) {
        return getDb().getDogDao().getAvailableDogs(ownerId);
    }

    public LiveData<List<Dog>> getOwnerDogs(long ownerId) {
        return getDb().getDogDao().getDogsbyOwner(ownerId);
    }


    public void removeOwnerFromDog(long ownerId, long dogId) {
        new Thread(() -> {
            getDb().getDogOwnerDao().delete(dogId, ownerId);
        }).start();
    }

    public LiveData<Dog> getDogById(long id) {
        return getDb().getDogDao().getDogById(id);
    }

    public Dog getDogByIdSync(long id) {
        return getDb().getDogDao().getDogByIdSync(id);
    }

    public void updateOwner(Owner owner) {
        new Thread(() ->
                getDb().getOwnerDao().updateOwner(owner)
        ).start();
    }

    public void updateOwnerSync(Owner owner) {
        getDb().getOwnerDao().updateOwner(owner);
    }

    public LiveData<Long> insertOwner(Owner owner) {
        MutableLiveData<Long> rowId = new MutableLiveData<>();
        Runnable r = () -> {
            long[] rowIds = getDb().getOwnerDao().insertOwner(owner);
            rowId.postValue(rowIds[0]);
        };
        new Thread(r).start();
        return rowId;
    }
    public long insertOwnerSync(Owner owner) {
        return getDb().getOwnerDao().insertOwner(owner)[0];
    }

    // Methods used in Dog Views

    public LiveData<List<Owner>> getAllAvailableOwners(long dogId) {
        return getDb().getOwnerDao().getAvailableOwners(dogId);
    }

    public LiveData<List<Owner>> getDogOwners(long dogId) {
        return getDb().getOwnerDao().getOwnersbyDog(dogId);
    }

    public void updateDogSync(Dog dog) {
        getDb().getDogDao().updateDog(dog);
    }

    public void addOwnerToDog(long ownerId, long dogId) {
        new Thread(() -> {
            getDb().getDogOwnerDao().insert(new DogOwner(dogId, ownerId));
        }).start();
    }

    public LiveData<Long> insertDog(Dog dog) {
        MutableLiveData<Long> rowId = new MutableLiveData<>();
        Runnable r = () -> {
            long[] rowIds = getDb().getDogDao().insertDog(dog);
            rowId.postValue(rowIds[0]);
        };
        new Thread(r).start();
        return rowId;
    }

    public long insertDogSync(Dog dog) {
        return getDb().getDogDao().insertDog(dog)[0];
    }

    public LiveData<Owner> getOwnerById(long id) {
        return getDb().getOwnerDao().getOwnerById(id);
    }

    public Owner getOwnerByIdSync(long id) {
        return getDb().getOwnerDao().getOwnerByIdSync(id);
    }

    // Methods used in Walk Views
    public LiveData<Long> insertWalkAndDogs(Walk walk, List<Dog> dogs) {

        MutableLiveData<Long> rowId = new MutableLiveData<>();
        Runnable r = () -> {
            long[] rowIds = getDb().getWalkDao().insertWalk(walk);
            long walkId = rowIds[0];
            List<WalkWithDogs> walkWithDogsList = new ArrayList<>();
            for (Dog dog : dogs) {
                WalkWithDogs walkWithDogs = new WalkWithDogs();
                walkWithDogs.setDogId(dog.getId());
                walkWithDogs.setWalkId(walkId);
                walkWithDogsList.add(walkWithDogs);
            }
            getDb().getWalkWithDogsDao().insert(walkWithDogsList);
            rowId.postValue(walkId);
        };
        new Thread(r).start();
        return rowId;
    }

    public LiveData<Walk> getWalkById(long walkId) {
        MutableLiveData<Walk> liveDataWalk = new MutableLiveData<>();
        new Thread(() -> {
            Log.e(TAG, "getWalkById: " + walkId);
            Walk walk = getDb().getWalkDao().getWalkById(walkId);
            walk.setDogs(getDb().getWalkWithDogsDao().getDogsOnWalk(walkId));
            liveDataWalk.postValue(walk);
        }).start();
        return liveDataWalk;
    }

    public void updateWalk(Walk walk) {
        new Thread(() ->
                getDb().getWalkDao().updateWalk(walk)
        ).start();
    }

    public LiveData<List<WalkLocation>> getLocations(long walkId) {
        return getDb().getWalkLocationDao().getLiveWalkLocations(walkId);
    }


    public void addPhotoToWalk(WalkPhoto walkPhoto) {
        new Thread(() ->
                getDb().getWalkPhotoDao().addPhoto(walkPhoto)
        ).start();
    }

    public void deleteWalkPhoto(long id) {
        new Thread(() ->
                getDb().getWalkPhotoDao().deletePhoto(id)
        ).start();
    }

    public LiveData<List<WalkPhoto>> getWalkPhotos(long walkId) {
        return getDb().getWalkPhotoDao().getAllWalkPhotos(walkId);
    }

    public LiveData<List<Owner>> getDogOwnersOnWalk(long walkId) {
        return getDb().getOwnerDao().getDogOwnersOnWalk(walkId);
    }

}

File: app/src/main/java/com/example/macarus0/walkiewalkie/viewmodel/WalkieViewModel.kt
package com.example.macarus0.walkiewalkie.viewmodel

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import android.util.Log

import com.example.macarus0.walkiewalkie.data.Dog
import com.example.macarus0.walkiewalkie.data.DogOwner
import com.example.macarus0.walkiewalkie.data.Owner
import com.example.macarus0.walkiewalkie.data.Walk
import com.example.macarus0.walkiewalkie.data.WalkLocation
import com.example.macarus0.walkiewalkie.data.WalkPhoto
import com.example.macarus0.walkiewalkie.data.WalkWithDogs
import com.example.macarus0.walkiewalkie.data.WalkieDatabase

import java.util.ArrayList

class WalkieViewModel(application: Application) : AndroidViewModel(application) {

    private val TAG = "WalkieViewModel"
    private val db = WalkieDatabase.getDb(application.applicationContext)

    // Methods used in overall views
    fun getAllDogs(): LiveData<List<Dog>> {
        return db.dogDao.getAllDogs()
    }

    val allOwners: LiveData<List<Owner>>
        get() = db.ownerDao.allOwners

    val allWalks: LiveData<List<Walk>>
        get() = db.walkDao.allWalks

    // Methods used in Owner Views

    fun getAllAvailableDogs(ownerId: Long): LiveData<List<Dog>> {
        return db.dogDao.getAvailableDogs(ownerId)
    }

    fun getOwnerDogs(ownerId: Long): LiveData<List<Dog>> {
        return db.dogDao.getDogsbyOwner(ownerId)
    }


    fun removeOwnerFromDog(ownerId: Long, dogId: Long) {
        Thread { db.dogOwnerDao.delete(dogId, ownerId) }.start()
    }

    fun getDogById(id: Long): LiveData<Dog> {
        return db.dogDao.getDogById(id)
    }

    fun updateOwnerSync(owner: Owner) {
        db.ownerDao.updateOwner(owner)
    }

    fun insertOwner(owner: Owner): LiveData<Long> {
        val rowId = MutableLiveData<Long>()
        val r = {
            val rowIds = db.ownerDao.insertOwner(owner)
            rowId.postValue(rowIds[0])
        }
        Thread(r).start()
        return rowId
    }

    fun insertOwnerSync(owner: Owner): Long {
        return db.ownerDao.insertOwner(owner)[0]
    }

    // Methods used in Dog Views

    fun getAllAvailableOwners(dogId: Long): LiveData<List<Owner>> {
        return db.ownerDao.getAvailableOwners(dogId)
    }

    fun getDogOwners(dogId: Long): LiveData<List<Owner>> {
        return db.ownerDao.getOwnersbyDog(dogId)
    }

    fun updateDogSync(dog: Dog) {
        db.dogDao.updateDog(dog)
    }

    fun addOwnerToDog(ownerId: Long, dogId: Long) {
        Thread { db.dogOwnerDao.insert(DogOwner(dogId, ownerId)) }.start()
    }

    fun insertDog(dog: Dog): LiveData<Long> {
        val rowId = MutableLiveData<Long>()
        val r = {
            val rowIds = db.dogDao.insertDog(dog)
            rowId.postValue(rowIds[0])
        }
        Thread(r).start()
        return rowId
    }

    fun insertDogSync(dog: Dog): Long {
        return db.dogDao.insertDog(dog)[0]
    }

    fun getOwnerById(id: Long): LiveData<Owner> {
        return db.ownerDao.getOwnerById(id)
    }

    // Methods used in Walk Views
    fun insertWalkAndDogs(walk: Walk, dogs: List<Dog>): LiveData<Long> {

        val rowId = MutableLiveData<Long>()
        val r = {
            val rowIds = db.walkDao.insertWalk(walk)
            val walkId = rowIds[0]
            val walkWithDogsList = ArrayList<WalkWithDogs>()
            for (dog in dogs) {
                val walkWithDogs = WalkWithDogs()
                walkWithDogs.dogId = dog.id
                walkWithDogs.walkId = walkId
                walkWithDogsList.add(walkWithDogs)
            }
            db.walkWithDogsDao.insert(walkWithDogsList)
            rowId.postValue(walkId)
        }
        Thread(r).start()
        return rowId
    }

    fun getWalkById(walkId: Long): LiveData<Walk> {
        val liveDataWalk = MutableLiveData<Walk>()
        Thread {
            Log.e(TAG, "getWalkById: $walkId")
            val walk = db.walkDao.getWalkById(walkId)
            walk.dogs = db.walkWithDogsDao.getDogsOnWalk(walkId)
            liveDataWalk.postValue(walk)
        }.start()
        return liveDataWalk
    }

    fun updateWalk(walk: Walk) {
        Thread { db.walkDao.updateWalk(walk) }.start()
    }

    fun getLocations(walkId: Long): LiveData<List<WalkLocation>> {
        return db.walkLocationDao.getLiveWalkLocations(walkId)
    }


    fun addPhotoToWalk(walkPhoto: WalkPhoto) {
        Thread { db.walkPhotoDao.addPhoto(walkPhoto) }.start()
    }

    fun deleteWalkPhoto(id: Long) {
        Thread { db.walkPhotoDao.deletePhoto(id) }.start()
    }

    fun getWalkPhotos(walkId: Long): LiveData<List<WalkPhoto>> {
        return db.walkPhotoDao.getAllWalkPhotos(walkId)
    }

    fun getDogOwnersOnWalk(walkId: Long): LiveData<List<Owner>> {
        return db.ownerDao.getDogOwnersOnWalk(walkId)
    }

}

